packages/gatsby/src/datastore/lmdb/query/create-index.ts

```diff
@@ -0,0 +1,304 @@
+    stats,
+  }
+
+  try {
+    let i = 0
+    for (const node of datastore.iterateNodesByType(typeName)) {
+      // Assuming materialization was run (executing custom resolvers for fields in `filter` and `sort` clauses)
+      //  And materialized values of those fields are stored in resolvedNodes
+      const resolvedFields = resolvedNodes?.get(node.id)
+      const { keys, multiKeyFields } = prepareIndexKeys(
+        node,
+        resolvedFields,
+        indexName,
+        indexFields
+      )
+      stats.keyCount += keys.length
+      stats.itemCount++
+      stats.maxKeysPerItem = Math.max(stats.maxKeysPerItem, keys.length)
+      indexMetadata.multiKeyFields.push(...multiKeyFields)
+
+      for (const indexKey of keys) {
+        // Note: this may throw if indexKey exceeds 1978 chars (lmdb limit) or contain objects/buffers/etc
+        indexes.put(indexKey, node.id)
+      }
+      if (++i % 5000 === 0) {
+        // Do not block event loop too much
+        await new Promise(resolve => setTimeout(resolve, 3))
+      }
+    }
+    indexMetadata.state = `ready`
+    indexMetadata.multiKeyFields = [...new Set(indexMetadata.multiKeyFields)]
+
+    await metadata.put(toMetadataKey(indexName), indexMetadata)
+    console.timeEnd(label)
+
+    return indexMetadata
+  } catch (e) {
+    indexMetadata.state = `error`
+    indexMetadata.error = String(e)
+    await metadata.put(toMetadataKey(indexName), indexMetadata)
+    throw e
+  }
+}
+
+/**
+ * Returns a list of index keys for a given node.
+ * One node may produce multiple index entries when indexing over array values.
+ *
+ * For example:
+ *  Node: { foo: [{ bar: `bar1`}, { bar: `bar2` }] }
+ *  Index fields: [`foo.bar`] will produce the following elements: [`bar1`, `bar2`]
+ *
+ * Keys are prefixed with index name and suffixed with node counter for stable sort.
+ *
+ * If materialization result (resolvedFields) exists for a given index field
+ *  it is used as a key element, otherwise the a raw node value is used.
+ */
+function prepareIndexKeys(
+  node: IGatsbyNode,
+  resolvedFields: { [field: string]: unknown } | undefined,
+  indexName: string,
+  indexFields: IndexFields
+): { keys: Array<IndexKey>; multiKeyFields: Array<string> } {
+  // TODO: use index id vs index name (shorter)
+  const indexKeyElements: Array<Array<IndexFieldValue>> = []
+  const multiKeyFields: Array<string> = []
+
+  indexKeyElements.push([indexName])
+  for (const dottedField of indexFields.keys()) {
+    const fieldValue = resolveFieldValue(dottedField, node, resolvedFields)
+    let indexFieldValue = jsValueToLmdbKey(fieldValue)
+
+    // Got value that can't be stored in lmdb key
+    if (typeof indexFieldValue === `undefined`) {
+      const path = `${node.internal.type}.${dottedField} (id: ${node.id})`
+      throw new Error(`Bad value at ${path}: ${inspect(fieldValue)}`)
+    }
+    indexFieldValue = Array.isArray(indexFieldValue)
+      ? indexFieldValue.flat() // FIXME
```

Comment:
can you expand on the `FIXME`? We shouldn't (have to) `.flat()` here?

Reply:
Good that I've put it there. The problem is with deeply nested arrays. The expected behavior is undefined. We don't have tests for those. I was going to add tests to see how our current implementation performs and then change here accordingly.

But obviously forgot. Will do. Thanks for spotting it 👍